File: Day74/Part5/boss_partition.py
def solve():
    try:
        line = input().split()
        if not line:
            return
        
        if len(line) == 1:
            n = int(line[0])
            line2 = input().split()
            d = [int(x) for x in line2]
            line3 = input().split()
            H = int(line3[0])
        else:
            if len(line) > 2:
                d = [int(x) for x in line]
                n = len(d)
                line2 = input().split()
                H = int(line2[0])
            else:
                n = int(line[0])
                line2 = input().split()
                d = [int(x) for x in line2]
                line3 = input().split()
                H = int(line3[0])
                
        # Subset sum
        sums = {0}
        for x in d:
            new_sums = set()
            for s in sums:
                new_sums.add(s + x)
            sums.update(new_sums)
            
        found = False
        for s in sums:
            if s > 0 and H % s == 0:
                found = True
                break
                
        print("1" if found else "0")
            
    except EOFError:
        pass

if __name__ == "__main__":
    solve()

File: Day74/Part5/boss_partition.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <set>

using namespace std;

int main() {
    int n;
    if (cin >> n) {
        vector<int> d(n);
        for (int i = 0; i < n; ++i) cin >> d[i];
        long long H;
        cin >> H;

        // Subset Sum problem.
        // Find if there exists a subset sum S such that H % S == 0.
        // Since N is likely small (standard CP constraint for subset sum is N <= 20 or sum is small).
        // If N is up to 20, 2^20 is 1M, feasible.
        // If N is larger, maybe H is small?
        // Assuming N is small enough for meet-in-the-middle or simple recursion if N <= 20.
        // Or bitset if sum is small.
        // Let's use a set for reachable sums.
        
        set<long long> sums;
        sums.insert(0);
        
        for (int x : d) {
            vector<long long> new_sums;
            for (long long s : sums) {
                new_sums.push_back(s + x);
            }
            for (long long s : new_sums) {
                sums.insert(s);
            }
        }
        
        bool found = false;
        for (long long s : sums) {
            if (s > 0 && H % s == 0) {
                found = true;
                break;
            }
        }
        
        cout << (found ? "1" : "0") << endl;
    }
    return 0;
}
